style/typescript/README.md

```diff
@@ -0,0 +1,58 @@
+# TypeScript
+
+[Sample TSX](sample.tsx) [ESLint Config](.eslintrc.json) [Prettier Config](.prettierrc)
+
+- Use the guidelines for [JavaScript](../javascript/README.md)
+- Use TypeScript in [strict mode]
+- Use [ESlint] and [Prettier] for auto-formatting and auto-fixing
+- Use [Jest](../testing-jest/README.md) for unit testing
+- Prefer [Functions] over [Classes]
+- Prefer [arrow functions] `=>`, over the `function` keyword except when using [Generics]
+- Use [Generic Constraints] when possible
+- Use `PascalCase` for [Interfaces] and [Type Aliases]
+- Use [readonly] properties where applicable
+- Use [Const Assertions] where applicable to avoid type widening
+- Avoid complex conditionals inside TSX
+- Avoid [Mixins]
+- Avoid [Decorators]
+- Avoid [Overloading Functions]
+- Prefer [async/await] over traditional promise syntax
+- Prefer [Optional Properties] in an interface rather than declaring the property type as `T | undefined`
+- Prefer explicitly defining interfaces over [Extending Interfaces]
+- Avoid the use of the [any] type
+- Avoid the [Non-null assertion operator]
+- Avoid [Type Assertions]
+- Prefer the `as`-syntax for [Type Assertions] over the angle-bracket syntax
+- Use the [Nullish coalescing operator]
+- Prefer [Type Guards] over [Type Assertions]
+- Prefer [Union Types], [Lookup Types], [Mapped Types] and [Const Assertions] over [Enums]
```

Comment:
I was talking about this with a client the other day, but how do you feel about preferring functions like `map`, `reduce`, `filter` as opposed to `for` loops and methods like `filterNulls(...)`? I'm not exactly sure what that category of functions is called...

Reply:
STRONGLY AGREE. What about something like

> Prefer functional style (`map`/`filter`/`reduce`) over iteration (`for`/`while`/`.each`).

Reply:
Love it.